### DISCUSSION TITLE
Capture keyboard events and suppress them

### DISCUSSION THREAD

**Discussion by NazarioJL** [Textualize/rich#1509] [Ideas] [answered]
It would be nice to capture keyboard events with the capability of handling such events to update a `rich` view model. 

For example, say I want to make an interactive display and capture input if I do the following:

```python
def make_panel(text: str = "Default value..."):
    text = Text(text)
    panel = Panel(
        text,
        padding=(1, 2),
        title=f"[b green]Enter a value to print, or 'Exit' to exit",
        border_style="bright_green",
    )

    return panel


if __name__ == '__main__':
    console = Console()
    main = Layout("main")  # this is effectively a view model
    main.update(make_panel())

    with Live(main, screen=False) as live:
        while True:
            string = console.input()  # pressing keys here makes the screen jitter, need enter to capture input
            if string.casefold() == "exit".casefold():
                break
            else:
                main.update(make_panel(string))
```
I would like to update my renderable based on some key input with a smooth output. Another example could be navigating an application with ⬆️➡️⬅️⬇️ keys for example.

**Comment by willmcgugan** [comment, marked as answer]
Rich doesn't have any support for handling keys beyond simple text entry. You may be interested in Rich's sister project, Textual.

**Reply by NazarioJL** [reply]
Seems like `Textual` is a superset of `Rich` then? It's like if life could not get any better. I'll look into `Textual`.

**Comment by codenotworking** [comment]
You should check out [`keyboard`](https://github.com/boppreh/keyboard) for capturing keyboard events.

Here's a very simple example:
```python
import keyboard

def pressed_key(keyevent: keyboard.KeyboardEvent):
    if keyevent.event_type == keyboard.KEY_DOWN:
        do_something()

# "suppress=True" suppresses keyboard events, so it doesn't echo to the terminal
keyboard.hook(pressed_key, suppress=True)
````

**Reply by NazarioJL** [reply]
Thanks @codenotworking I had tried this already. Did not get the results I was looking for.
